homeassistant/components/lutron_caseta/__init__.py

```diff
@@ -218,57 +221,151 @@ def _async_register_bridge_device(
     hass: HomeAssistant,
     config_entry_id: str,
     bridge: Smartbridge,
     bridge_device: dict[str, Any],
-    button_devices_by_id: dict[int, dict],
-) -> tuple[dict[str, dict], dict[int, DeviceInfo]]:
-    """Register button devices (Pico Remotes) in the device registry."""
+) -> LutronKeypadData:
+    """Register keypad devices (Keypads and Pico Remotes) in the device registry."""
+
     device_registry = dr.async_get(hass)
-    button_devices_by_dr_id: dict[str, dict] = {}
-    device_info_by_device_id: dict[int, DeviceInfo] = {}
-    seen: set[str] = set()
+
     bridge_devices = bridge.get_devices()
+    bridge_buttons = bridge.buttons
+
+    dr_device_id_to_keypad: dict[str, dict] = {}
+    keypads: dict[int, Any] = {}
+    keypad_buttons: dict[int, Any] = {}
+    keypad_button_names_to_leap: dict[int, dict[str, int]] = {}
+
+    keypad: dict[str, Any] = {}
+    for bridge_button in bridge_buttons.values():
+
+        bridge_keypad = bridge_devices[bridge_button["parent_device"]]
 
-    for device in button_devices_by_id.values():
+        if bridge_keypad["device_id"] not in keypads:
+            # First time seeing this keypad, build keypad data and store in keypads
+            area_name = _area_name_from_id(bridge.areas, bridge_keypad["area"])
+            keypad_name = bridge_keypad["name"].split("_")[-1]
+            keypad_serial = _handle_none_keypad_serial(
+                bridge_keypad, bridge_device["serial"]
+            )
+            device_args: DeviceInfo = {
+                "name": f"{area_name} {keypad_name}",
+                "manufacturer": MANUFACTURER,
+                "identifiers": {(DOMAIN, keypad_serial)},
+                "model": f"{bridge_keypad['model']} ({bridge_keypad['type']})",
+                "via_device": (DOMAIN, bridge_device["serial"]),
+            }
+            if area_name != UNASSIGNED_AREA:
+                device_args["suggested_area"] = area_name
+
+            dr_device = device_registry.async_get_or_create(
+                **device_args, config_entry_id=config_entry_id
+            )
 
-        ha_device = device
-        if "parent_device" in device and device["parent_device"] is not None:
-            # Device is a child of parent_device
-            # use the parent_device for HA device info
-            ha_device = bridge_devices[device["parent_device"]]
+            keypad = keypads.setdefault(
+                bridge_keypad["device_id"],
+                {
+                    "lutron_device_id": bridge_keypad["device_id"],
+                    "dr_device_id": dr_device.id,
+                    "area_id": bridge_keypad["area"],
+                    "area_name": area_name,
+                    "name": keypad_name,
+                    "serial": keypad_serial,
+                    "device_info": device_args,
+                    "model": bridge_keypad["model"],
+                    "type": bridge_keypad["type"],
+                    "buttons": [],
+                },
```

Comment:
Let's make this a typed dict so we don't risk mistakes on types

Reply:
Suggested change (not tested)

https://github.com/bdraco/home-assistant/commit/5cff38c4ab2e97b370e897e2dfec7f28e15aabec

Reply:
I've implemented the Typed Dict